[SPARK-8979] Add a PID based rate estimator

dragos opened the pull request:
Based on #7600 

/cc @tdas

huitseeker committed:
[SPARK-8979][Streaming] Implements a PIDRateEstimator

dragos committed:
Update to latest version of the code, various style and name improvements.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38377 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38377/consoleFull) for   PR 7648 at commit [`70d875b`](https://github.com/apache/spark/commit/70d875b14c944ac2aab1b5f8dfad06f2e6b281b8).

SparkQA commented:
[Test build #38377 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38377/console) for   PR 7648 at commit [`70d875b`](https://github.com/apache/spark/commit/70d875b14c944ac2aab1b5f8dfad06f2e6b281b8).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38381 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38381/consoleFull) for   PR 7648 at commit [`2cd7139`](https://github.com/apache/spark/commit/2cd713900dd8b81657a0f8d11a90202f081da22d).

SparkQA commented:
[Test build #38381 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38381/console) for   PR 7648 at commit [`2cd7139`](https://github.com/apache/spark/commit/2cd713900dd8b81657a0f8d11a90202f081da22d).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,91 @@
+ * the speed of ingestion of elements into Spark Streaming.
+ *
+ * @param batchDurationMillis the batch duration, in milliseconds
+ * @param proportional how much the correction should depend on the current
```
@dragos @huitseeker Could you change the documentation to make it easier to understand in the context of streaming? For example, its not clear what is "error" in the context of streaming. So these docs needs to be expanded to explain how the processing and scheduling delay is used to estimate the rate, and how the parameters proportional and integral controls the processing. And what are recommended ranges for these parameters, and how do you control them.

dragos replied:
Currently they are not user-configurable, should we do that? I think it's too much control for most people, but on the other hand it's the only knobs we have, short of writing another rate estimator altogether.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,91 @@
+      schedulingDelay: Long // in milliseconds
+      ): Option[Double] = {
+
+    this.synchronized {
```
Is there any reason for this synchronized? Do you think this will be called from multiple threads?

dragos replied:
This may be called by different threads, either due to the streaming listener bus being asynchronous, or the rate controller using a future when calling this method. Probably it's unlikely to have concurrent runs, but at least for visibility of updates we need some form of synchronization. We could remove this and make all fields `volatile`, but I'm not 100% clear if there's really no chance of running concurrently in some degenerate situation.

dragos commented:
I pushed an expanded doc, but it would be good if @huitseeker had a look too, since this is mainly his brainchild.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build finished. Test FAILed.

tdas commented:
Jenkins, test this please.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #38861 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38861/consoleFull) for   PR 7648 at commit [`491e371`](https://github.com/apache/spark/commit/491e3712ce5b75f09919e865e7b3855b5cb2febd).

SparkQA commented:
[Test build #38861 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/38861/console) for   PR 7648 at commit [`491e371`](https://github.com/apache/spark/commit/491e3712ce5b75f09919e865e7b3855b5cb2febd).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented:
@dragos Please update this PR with latest master. Would help to review this.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+/**
+ * Implements a proportional-integral-derivative (PID) controller which acts on
+ * the speed of ingestion of elements into Spark Streaming. A PID controller works
+ * by calculating an '''error''' between a measured output and a desired setpoint. In the
```
What is the `set point`?

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+      elements: Long,
+      processingDelay: Long, // in milliseconds
+      schedulingDelay: Long // in milliseconds
+      ): Option[Double] = {
```
two space on this line `)`

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        // in elements/second
+        val error = latestRate - processingSpeed
+
+        // in elements/second
```
Can you write down the physical interpretation of this sumError? cc @huitseeker

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        val delaySinceUpdate = (time - latestTime).toDouble / 1000
+
+        // in elements/second
+        val processingSpeed = elements.toDouble / processingDelay * 1000
```
Please use the consistent naming. speed --> rate

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        val processingSpeed = elements.toDouble / processingDelay * 1000
+
+        // in elements/second
+        val error = latestRate - processingSpeed
```
Could you make the names more semantically meaningful? How about: error --> changeInRate?

tdas replied:
Why is the latestRate considered as the set point (that's my assumption since the error is calculated between the observed value and the set point, according to PID theory)? @huitseeker

dragos replied:
Since @huitseeker seems to be away, I'll answer this.

The `latestRate` is what we considered the desired value at the previous batch update. With the new information we got for the last batch interval, we compute a current rate, and compare to what we asked for, that's constitutes our error that needs correction.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        val error = latestRate - processingSpeed
+
+        // in elements/second
+        val sumError = schedulingDelay.toDouble * processingSpeed / batchIntervalMillis
```
Its hard to understand what sumError mean in terms of the rates and all? Can you write down the physical interpretation of this sumError? And also make the name better accordingly? 
cc @huitseeker

tdas replied:
So I am trying to understand this. 
(scheduling delay / batch interval) = approx the number of batches the system is delayed. Lets call it numDelayedBatches.
Now you are multiplying numDelayedBatches X processingSpeed. So you are scaling the current processing rate with number of batches that are delayed. Right?

tdas commented:
@dragos Is there any explanation/derivation of the implementation from the PID concepts written up somewhere. Would be a great idea if that is attached to the JIRA.

dragos commented:
@huitseeker wrote one, he can add it himself, it wasn't on Google Docs, but a local document that I can't find right now.

dragos committed:
A couple of variable renames.

AmplabJenkins commented:
Merged build triggered.

dragos commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        val processingRate = elements.toDouble / processingDelay * 1000
+
+        // in elements/second
+        val error = latestRate - processingRate
```
Carrying over conversation from previous thread that got lost due to rebase

>Could you make the names more semantically meaningful? How about: error --> changeInRate?
@tdas
>tdas added a note 14 hours ago
>Why is the latestRate considered as the set point (that's my assumption since the error is calculated between the observed value and the set point, according to PID theory)? @huitseeker
@dragos   
dragos added a note 2 hours ago
>Since @huitseeker seems to be away, I'll answer this.

>The latestRate is what we considered the desired value at the previous batch update. With the new information we got for the last batch interval, we compute a current rate, and compare to what we asked for, that's constitutes our error that needs correction.

dragos replied:
Here I'd prefer to keep this as `error`, as I think most people reading this code would have more troubles mapping things to PID terminology than to Spark Streaming terminology, and all PID docs will mention error and correction.

tdas replied:
But that is exactly the problem, people who are reading the streaming code (like me) is more like to know streaming than PID concepts, and if the code does not make it clear in terms of streaming, it is super hard to relate to. So I am fine with keep this as `error` as long as there is a explanation in terms of streaming stuff along with it. Just like what you added for historicalError.

dragos replied:
I'll add the explanation.

dragos commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        val error = latestRate - processingRate
+
+        // in elements/second
+        val sumError = schedulingDelay.toDouble * processingRate / batchIntervalMillis
```
Carrying over conversation from previous thread that got lost due to rebase

>Its hard to understand what sumError mean in terms of the rates and all? Can you write down the physical interpretation of this sumError? And also make the name better accordingly? 
cc @huitseeker
@tdas
tdas added a note 14 hours ago
So I am trying to understand this. 
(scheduling delay / batch interval) = approx the number of batches the system is delayed. Lets call it numDelayedBatches.
Now you are multiplying numDelayedBatches X processingSpeed. So you are scaling the current processing rate with number of batches that are delayed. Right?

YikSanChan replied:
> from the number of overflowing elements we can calculate the rate at which they would be cleared by dividing it by the batch interval

@dragos Why? I still don't understand why historical error = # overflowing elements / batch interval. The explanation that if # overflowing elements = 0, then historical error = 0 does not explain the 1 / batch interval coefficient.

I feel the hard part is: It is easy to reason about the accumulated error if error is an absolute number, for example, the number of overflowing elements, but what is an accumulated error of a rate?

Setting up 1 / batch interval as the coefficient may not be a big problem anyway, because at the end of the day it is nothing but a constant that you can tune the integral coefficient to compensate.

dragos replied:
@YikSanChan sorry for the delay, it's been a while since I didn't visit this code. I admit I am not sure I understand where you're coming from, and I don't want to simply repeat what I wrote above. So I'll ask a question or two myself first:

What prompted you to look into this code? Do you see bad behavior on certain loads?

> The explanation that if # overflowing elements = 0, then historical error = 0 does not explain the 1 / batch interval coefficient.

If the overflowing elements are 0, doesn't this imply the `schedulingDelay` is zero, so this whole term is 0 as well (instead of `1/batchInterval`)?

>I feel the hard part is: It is easy to reason about the accumulated error if error is an absolute number, for example, the number of overflowing elements, but what is an accumulated error of a rate?

If we agree that the accumulated error is what it is computed above as number of elements, then "all we need to do" is to transform it into a rate (elements/second). We need to do that since that's the quantity that this PID is controlling. By dividing it by the batch interval we get how many `elements/second` are overflowing, in other words, by how much is the existing rate "off", if we wanted to have no overflowing elements. Does this make sense to you?

YikSanChan replied:
@dragos Thanks for responding!

> What prompted you to look into this code? Do you see bad behavior on certain loads?

I am not seeing bad behavior because I am not using spark streaming. The reason why I look into the code is I try to leverage the PID controller in an imaginary scenario I describe [here](https://robotics.stackexchange.com/questions/20248/how-to-determine-actuator-and-process-for-pid-and-how-to-intepret-the-output). I use Scala and Akka so I look for PID controller example in Scala on GitHub, which brings me here.

Your implementation gives me a nice picture of how to turn a "rule of thumb" (that's how I feel about PID) into an actual implementation that is well explained and understood (with clear physical/system interpretation) and runs well in production. Thank you :)

> If the overflowing elements are 0, doesn't this imply the schedulingDelay is zero, so this whole term is 0 as well (instead of 1/batchInterval)?

Yes, you are right. But what I am asking is: in the equation `historicalError = schedulingDelay * processingRate / batchInterval`, why is the coefficient `1 / batchInterval`, rather than `1 / 2`, `1 / 100`, `1 / sum(batchInteval)`, or whatever else?

> in other words, by how much is the existing rate "off", if we wanted to have no overflowing elements.

If I understand correctly, the overflowing elements are accumulated since batch 0 (provided we're at batch n). If we all agree with the statement, then shouldn't the rate be `1 /(n * batchInterval)` instead?

dragos committed:
Add configuration for PID.

AmplabJenkins commented:
Merged build triggered.

dragos commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,100 @@
+        val error = latestRate - processingRate
+
+        // in elements/second
+        val sumError = schedulingDelay.toDouble * processingRate / batchIntervalMillis
```
Here's the gist of it:

- we consider `schedulingDelay` as an indication of accumulated error, which corresponds to the integral part in a PID controller. Intuitively it makes sense: the fact that there is a delay means we had too many elements in previous batches, and the system can't process them in the given batch interval

The challenge is to transform this indication from *time* to a rate, which is the quantity that our PID is measuring (and controlling). Here's the reasoning:

- a scheduling delay `s` corresponds to `s * processingRate` *overflowing* elements. Those are elements that couldn't be processed in previous batches, leading to this delay. We assume the processingRate didn't change too much (since it's mostly a measure of the cluster performance, with small variations like checkpointing), but a good approximation
-  from the number of overflowing elements we can calculate the rate at which they would be cleared by dividing it by the batch interval. This rate is our "historical" error, or integral part, since if we subtracted this rate from the previous "calculated rate", there wouldn't have been any overflowing elements, and the scheduling delay would have been zero.

There's some additional details about units of measure, since schedulingDelay and the batchInterval are in milliseconds by rates are in elements/second, but if you do the math you'd notice that the 1000s cancel out.

dragos committed:
Better explanation of historicalError.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build started.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39054 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39054/consoleFull) for   PR 7648 at commit [`93b74f8`](https://github.com/apache/spark/commit/93b74f884ea17da65297e47ff9a20b53d93225d1).

SparkQA commented:
[Test build #39052 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39052/consoleFull) for   PR 7648 at commit [`7975b0c`](https://github.com/apache/spark/commit/7975b0c9703696653563d2b457b4ef071f30bfe9).

SparkQA commented:
[Test build #39047 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39047/consoleFull) for   PR 7648 at commit [`26cfd78`](https://github.com/apache/spark/commit/26cfd78c339e58e71c138e424952002f13595389).

SparkQA commented:
[Test build #39054 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39054/console) for   PR 7648 at commit [`93b74f8`](https://github.com/apache/spark/commit/93b74f884ea17da65297e47ff9a20b53d93225d1).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public static final class FloatPrefixComparator extends PrefixComparator `
  * `case class UnsafeExternalSort(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #39052 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39052/console) for   PR 7648 at commit [`7975b0c`](https://github.com/apache/spark/commit/7975b0c9703696653563d2b457b4ef071f30bfe9).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

SparkQA commented:
[Test build #39047 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39047/console) for   PR 7648 at commit [`26cfd78`](https://github.com/apache/spark/commit/26cfd78c339e58e71c138e424952002f13595389).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `  public static final class FloatPrefixComparator extends PrefixComparator `
  * `case class UnsafeExternalSort(`

AmplabJenkins commented:
Merged build finished. Test PASSed.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/RateEstimator.scala:
```diff
@@ -48,12 +49,21 @@ object RateEstimator {
+  def create(conf: SparkConf, batchInterval: Duration): RateEstimator =
+    conf.get("spark.streaming.backpressure.rateEstimator", "pid") match {
+      case "pid" =>
+        val proportional = conf.getDouble("spark.streraming.backpressure.pid.proportional", -1.0)
```
typo!

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,109 @@
+ */
+private[streaming] class PIDRateEstimator(
+    batchIntervalMillis: Long,
+    proportional: Double = -1D,
```
Let's actually convert all the parameters to be all positive. Would be more intuitive to say that the proportional parameter can be set between 0 - 1 -- 1 = max aggressive and 0 = least aggressive

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,109 @@
+    s"Specified batch interval $batchIntervalMillis in PIDRateEstimator is invalid.")
+
+  def compute(time: Long, // in milliseconds
+      elements: Long,
```
elements --> numElements

dragos replied:
I don't really like this convention. We already have types in Scala, but fine.

tdas replied:
Well the goals is to stay consistent with the surrounding codebase you are contributing to, isnt it? 
We have made it clear that we do not always use Scala convention and rather have our own style guide that we think makes the code easier to read and understand, even by non-scala folks.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,109 @@
+        // a scheduling delay s corresponds to s * processingRate overflowing elements. Those
+        // are elements that couldn't be processed in previous batches, leading to this delay.
+        // We assume the processingRate didn't change too much.
+        // from the number of overflowing elements we can calculate the rate at which they would be
```
This is good, but there are grammatical issues. Please revise. E.g. "from" should be capitalized if it was to be start of a sentence.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,109 @@
+ * @param proportional how much the correction should depend on the current
+ *        error. This term usually provides the bulk of correction. A value too large would
+ *        make the controller overshoot the setpoint, while a small value would make the
+ *        controller too insensitive. The default value is -1.
```
Please define the range, along with the explanation. (assuming with make it positive) the acceptable range is 0 - 1 where 1 is most aggressive to updating the current rate, and 0 least aggressive. Also the range should be verified, for all the parameters.

dragos replied:
I don't think we should limit the range to [0, 1]. There is no such limit in a typical PID controller, and I'm not sure we stand much to gain. These are just constants. We could require them to be strictly positive and modify the formula, but we're departing from PID theory. Presumably that makes sense in our domain.

tdas replied:
Assuming that we are switching the signs (i.e. default is 1 instead of -1), what does proportional < 0 mean?
That is just plain wrong! You are increasing the rate when it should be decreased. And anything more than 1 means that you are compensating more than feasible. Consider this. say there is a workload with 1 second batches where the system can process 5 records in 0.8 seconds. Say the rate limit was 10. After first batch with 5 recods in 0.8 seconds process delay (no scheduling delay), the calculation would be (assuming proportional = 1.5)
```
error = rateLimit - numRecords/procDelay 10 - (5/0.8) = 10 - 6.25 = 3.75
newRate = rateLimit - 1.5 * error = 10 - 1.5 * 3.75 = 4.375 
```
4.375 is less than 5 records that the system can handle in 0.8 secs. This is wrong, that is going to happen for proportional > 1. With proportional = 1, the calculation is.
`new rate = 10 - 1.0 * (10 - (5/0.8)) = 6.25  `
This same as it had processed in the previous batch. Perfectly feasible.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimatorSuite.scala:
```diff
@@ -0,0 +1,101 @@
+
+class PIDRateEstimatorSuite extends SparkFunSuite with Matchers {
+
+  test("first bound is None") {
```
please add an unit test that is going to verify whether setting the appropriate params in the SparkConf create the right estimator and the right estimator parameters.

tdas replied:
Also add unit test to check whether setting the parameters to incorrect ranges throws errors or not (use intercept[ ] to catch errors)

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimatorSuite.scala:
```diff
@@ -0,0 +1,101 @@
+
+  test("works even with no time between updates") {
+    val p = new PIDRateEstimator(20, -1D, 0D, 0D)
+    p.compute(0, 10, 10, 0)
```
This is hard to understand. Basically the internal state is being mutated in these series of calls. So could you put asserts on the internal state so that I can understand what are the expected changes

dragos replied:
This test isn't really about mutating state, but checking that two consecutive updates for the same time doesn't trip the estimator (for instance, a division by zero).

tdas replied:
Then it will easier to get if inline comments said so ..  // same time as previous update

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimatorSuite.scala:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  test("works even with no time between updates") {
+    val p = new PIDRateEstimator(20, -1D, 0D, 0D)
```
If the same parameters are used everywhere, please put this in a function. Easier to read.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,109 @@
+  private var latestRate: Double = -1D
+  private var latestError: Double = -1L
+
+  require(
```
Should verify the values of the parameters.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimatorSuite.scala:
```diff
@@ -0,0 +1,101 @@
+  }
+
+  test("bound is never negative") {
+    val p = new PIDRateEstimator(20, -1D, -1D, 0D)
```
For each of the test could you explain in words the kind of input pattern you are testing against, and what the expected output be? Its kinda hard to understand it from the tests itself, so I am not to verify whether the test is doing what its supposed to do.

tdas replied:
This is important, please add this, otherwise its impossible for me to review and is a blocker.

dragos committed:
Add more comments, made all PID constant parameters positive, a couple more tests.

dragos commented:
I think I addressed all comments. In case I missed something, @nraychaudhuri will take over next week. I'm travelling today and won't be able to make further changes, but it's my firm belief this is in pretty good shape now. I hope it can get in.

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #39203 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39203/consoleFull) for   PR 7648 at commit [`aa5b097`](https://github.com/apache/spark/commit/aa5b097e3d58f07feb830f0ff5456088064b7fff).

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,124 @@
+    s"Specified batch interval $batchIntervalMillis in PIDRateEstimator is invalid.")
+  require(
+    proportional >= 0,
+    s"Proportional term $proportional in PIDRateEstimator should be >= 0.")
```
Did you see my earlier comment about > 1 not being practically feasible.

tdas commented on streaming/src/test/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimatorSuite.scala:
```diff
@@ -0,0 +1,137 @@
+    conf.set("spark.streaming.backpressure.rateEstimator", "pid")
+    val pid = RateEstimator.create(conf, Seconds(1))
+    pid.getClass should equal(classOf[PIDRateEstimator])
+  }
```
Test for the `spark.streaming.backpressure.rateEstimator.pid.*` params as well.

tdas commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/rate/PIDRateEstimator.scala:
```diff
@@ -0,0 +1,124 @@
+ */
+private[streaming] class PIDRateEstimator(
+    batchIntervalMillis: Long,
+    proportional: Double = 1D,
```
nit: I dont particularly mind, but is there any particular reason for having D at the end everywhere. Looks pretty weird.

SparkQA commented:
[Test build #39203 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/39203/console) for   PR 7648 at commit [`aa5b097`](https://github.com/apache/spark/commit/aa5b097e3d58f07feb830f0ff5456088064b7fff).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

nraychaudhuri commented:
@tdas I will try to take care of this by early next week.

tdas commented:
Since we are nearing feature freeze today, I am going to merge this in. There are few nit, but none of them are blockers and we can figure out those details next week.

tdas commented:
I am merging this. Thanks @dragos and @huitseeker
